rbc/heavydb/column_array.py

```diff
@@ -0,0 +1,352 @@
+"""
+
+__all__ = ['HeavyDBOutputColumnArrayType', 'HeavyDBColumnArrayType', 'ColumnArray']
+
+import operator
+from typing import TypeVar
+from rbc import typesystem
+from .column import HeavyDBColumnType
+from . import array
+from .metatype import HeavyDBMetaType
+from numba.core import extending, cgutils
+from numba.core import types as nb_types
+from llvmlite import ir
+
+
+T = TypeVar('T')
+
+
+i1 = ir.IntType(1)
+i8 = ir.IntType(8)
+i8p = i8.as_pointer()
+i8pp = i8p.as_pointer()
+i64 = ir.IntType(64)
+i64p = i64.as_pointer()
+void = ir.VoidType()
+
+
+_COLUMN_PARAM_NAME = 'ColumnArray_inner_type'
+
+
+class ColumnArray(metaclass=HeavyDBMetaType):
+    """
+    RBC ``Column<Array<T>>`` type that corresponds to HeavyDB COLUMN<ARRAY<T>>
+
+    In HeavyDB, a Column of type ``Array<T>`` is represented as follows:
+
+    .. code-block:: c
+
+        {
+            int8_t* flatbuffer;
+            int64_t size;
+        }
+
+    """
+    def get_item(self, index: int) -> int:
+        """
+        Return ``self[index]``
+
+        .. note::
+            Only available on ``CPU``
+        """
+
+    def set_item(self, index: int, arr: 'array.Array[T]') -> int:
+        """
+        ``self[index] = arr``
+
+        .. note::
+            Only available on ``CPU``
+        """
+
+    def concat_item(self, index: int, arr: 'array.Array[T]') -> int:
+        """
+        Concat array ``arr`` at ``index``.
+
+        .. note::
+            Only available on ``CPU``
+        """
+
+    def is_null(self, index: int) -> int:
+        """
+        Check if the array at ``index`` is null.
+
+        .. note::
+            Only available on ``CPU``
+        """
+
+    def set_null(self, index: int) -> int:
+        """
+        Set the array at ``index`` to null.
+
+        .. note::
+            Only available on ``CPU``
+        """
+
+    def __len__(self) -> int:
+        """
+        """
+
+
+class HeavyDBColumnArrayType(HeavyDBColumnType):
+
+    def postprocess_type(self):
+        return self
+
+    @property
+    def numba_type(self):
+        return ColumnArrayType
+
+    @property
+    def numba_pointer_type(self):
+        return ColumnArrayPointer
+
+    @property
+    def element_type(self):
+        return typesystem.Type.fromstring('int8_t')
+
+    @property
+    def custom_params(self):
+        return {
+            **super().custom_params,
+            _COLUMN_PARAM_NAME: self[0][0],
+            'name': f'STRUCT_{self.mangling()}_{self[0][0][0][0].tostring()}'
+        }
+
+
+class HeavyDBOutputColumnArrayType(HeavyDBColumnArrayType):
+    """Heavydb OutputColumn type for RBC typesystem.
+
+    OutputColumn is the same as Column<Array<T>> but introduced to distinguish
+    the input and output arguments of UDTFs.
+    """
+
+
+class ColumnArrayType(nb_types.Type):
+    """Numba type class for HeavyDB buffer structures.
+    """
+
+    def __init__(self, dtype):
+        array_T = self.__typesystem_type__._params.get(_COLUMN_PARAM_NAME)
```

Comment:
IIUC, `self` is `numba.core.types.Type` derived instance. Where is defined its `__typesystem_type__` attribute? In numba?

Reply:
Here:
https://github.com/xnd-project/rbc/blob/c576db2714c950b31bf799590d91bd42df768b57/rbc/structure_type.py#L93